libcxx/include/__flat_map/flat_map.h

```diff
@@ -0,0 +1,1280 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _LIBCPP___FLAT_MAP_FLAT_MAP_H
+#define _LIBCPP___FLAT_MAP_FLAT_MAP_H
+
+#include <__algorithm/lexicographical_compare_three_way.h>
+#include <__algorithm/ranges_equal.h>
+#include <__algorithm/ranges_lower_bound.h>
+#include <__algorithm/ranges_partition_point.h>
+#include <__algorithm/ranges_stable_sort.h>
+#include <__algorithm/ranges_unique.h>
+#include <__algorithm/ranges_upper_bound.h>
+#include <__compare/synth_three_way.h>
+#include <__concepts/convertible_to.h>
+#include <__config>
+#include <__flat_map/container_traits.h>
+#include <__flat_map/sorted_unique.h>
+#include <__functional/invoke.h>
+#include <__functional/is_transparent.h>
+#include <__functional/operations.h>
+#include <__iterator/concepts.h>
+#include <__iterator/distance.h>
+#include <__iterator/iterator_traits.h>
+#include <__iterator/ranges_iterator_traits.h>
+#include <__iterator/reverse_iterator.h>
+#include <__memory/allocator_traits.h>
+#include <__memory/uses_allocator.h>
+#include <__memory/uses_allocator_construction.h>
+#include <__ranges/concepts.h>
+#include <__ranges/container_compatible_range.h>
+#include <__ranges/ref_view.h>
+#include <__ranges/subrange.h>
+#include <__ranges/zip_view.h>
+#include <__type_traits/conjunction.h>
+#include <__type_traits/invoke.h>
+#include <__type_traits/is_allocator.h>
+#include <__type_traits/is_nothrow_constructible.h>
+#include <__type_traits/maybe_const.h>
+#include <__utility/pair.h>
+#include <initializer_list>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+#if !defined(_LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER)
+#  pragma GCC system_header
+#endif
+
+#if _LIBCPP_STD_VER >= 23
+
+_LIBCPP_BEGIN_NAMESPACE_STD
+
+template <class _Key,
+          class _Tp,
+          class _Compare         = less<_Key>,
+          class _KeyContainer    = vector<_Key>,
+          class _MappedContainer = vector<_Tp>>
+class flat_map {
+  template <bool _Const>
+  struct __iterator;
+
+  template <class, class, class, class, class>
+  friend class flat_map;
+
+public:
+  // types
+  using key_type    = _Key;
+  using mapped_type = _Tp;
+  using value_type  = pair<key_type, mapped_type>;
+  using key_compare = _Compare;
+  // TODO : the following is the spec, but not implementable for vector<bool>
```

Comment:
Interestingly, the spec http://eel.is/c++draft/flat.map#overview-note-3 explicitly says vector<bool> is not a sequence container, which means that,
`std::flat_map<int, bool>` is not well formed.

Reply:
That's interesting (and kinda broken)! I would suggest adding a `static_assert` to make sure we don't allow users to use `flat_map<*, bool>` or `flat_map<bool, *>` until this has been resolved in the Standard. Otherwise, we expose ourselves to Hyrum's law and it being really difficult to implement the actual resolution of a LWG issue targeting this.

So basically, I would file a LWG issue and disallow this entirely for now.

Reply:
draft:

# Ttile
`flat_map<*, bool>` or `flat_map<bool, *>`  is ill-formed

# Content
`flat_map` requires the underlying container to meet the "sequence container" requirement and `vector<bool>` is not a "sequence container"
http://eel.is/c++draft/flat.map#overview-7

As a result, `flat_map<*, bool>` or `flat_map<bool, *>`  is ill-formed. This would be surprising to the users. In fact, supporting `vector<bool>` is not difficult at all. In libc++, all we need to do is to change
```
- using reference = pair<const key_type&, mapped_type&>;
+ using reference = pair<range_reference_t<const KeyContainer>, range_reference_t<MappedContainer>>;
```
and `flat_map<*, bool>` or `flat_map<bool, *>` works fine.

# Proposed Wording
None

Reply:
> Given this note (http://eel.is/c++draft/flat.map#overview-note-3):
> 
> > Note [3](http://eel.is/c++draft/flat.map#overview-note-3): vector is not a sequence container[.](http://eel.is/c++draft/flat.map#overview-7.sentence-3) — end note]
> 
> I think this has been discussed already and it looks like the spec is non-ambiguous that `flat_map<bool, *>` isn't supported.

Per https://github.com/tzlaine/flat_map/commit/bb74f2bb3dc7e37d3c7332ac7367521fbf925972, this was probably discussed in the Kona 2019 meeting.